Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: b9ca77ca5730a4d6a15889198e29dc9ac82465a5
Number of Error(s): 5
CMake Error at cmake/Modules/DownloadLcbDep.cmake:18 (MESSAGE):
-- Configuring incomplete, errors occurred!
make[2]: *** [plugins/couchbase/CMakeFiles/libcouchbase-build.dir/build.make:62: plugins/couchbase/libcouchbase-build/lib/libcouchbase.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:15764: plugins/couchbase/CMakeFiles/libcouchbase-build.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
CMake Error at cmake/Modules/DownloadLcbDep.cmake:18 (MESSAGE):
  Failed to download
  https://packages.couchbase.com/clients/c/mock/CouchbaseMock-1.5.19.jar
Call Stack (most recent call first):
  tests/CMakeLists.txt:70 (DOWNLOAD_LCB_DEP)
-- Configuring incomplete, errors occurred!
See also /mnt/disk1/home/vamosax/smoketest/PR-13091/build/plugins/couchbase/libcouchbase-build/CMakeFiles/CMakeOutput.log.
Checking librdkafka integrity
librdkafka.so.1                OK
librdkafka.a                   OK
make[2]: *** [plugins/couchbase/CMakeFiles/libcouchbase-build.dir/build.make:62: plugins/couchbase/libcouchbase-build/lib/libcouchbase.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:15764: plugins/couchbase/CMakeFiles/libcouchbase-build.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
Symbol visibility              OK
Checking librdkafka++ integrity
librdkafka++.so.1              OK
librdkafka++.a                 OK
Install librdkafka to /mnt/disk1/home/vamosax/smoketest/PR-13091/build/plugins/kafka
Install librdkafka++ to /mnt/disk1/home/vamosax/smoketest/PR-13091/build/plugins/kafka
[ 20%] Built target librdkafka-build
configure.ac:35: installing ./ar-lib
configure.ac:26: installing ./compile
configure.ac:22: installing ./config.guess
configure.ac:22: installing ./config.sub
configure.ac:21: installing ./install-sh
configure.ac:21: installing ./missing
Makefile.am: installing ./depcomp
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for style of include used by make... GNU
checking dependency style of gcc... gcc3
checking whether gcc and cc understand -c and -o together... yes
checking if gcc supports -fvisibility=hidden flag... yes
checking if gcc supports -g flag... yes
checking if gcc supports -std=gnu89 flag... yes
checking if gcc supports -pedantic flag... yes
checking if gcc supports -Wall flag... yes
checking if gcc supports -Wextra flag... yes
checking if gcc supports -Wno-unused-parameter flag... yes
checking for ar... ar
checking the archiver (ar) interface... ar
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld
checking if the linker (/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld) 
 ... ( comment is too long, 361 lines, 27183 bytes, truncated) ... 
[Build error:](https://gist.github.com/c8a79e798832e600b156c400cf3fe5cb)
Build log [Build log](https://gist.github.com/HPCCSmoketest/c8a79e798832e600b156c400cf3fe5cb#file-result-19-11-13-10-11-40-log)

```

--- a/esp/scm/ws_smc.ecm
+++ b/esp/scm/ws_smc.ecm
@@ -157,6 +157,8 @@ ESPresponse [exceptions_inline] ActivityResponse
     [min_ver("1.12")] bool SuperUser(false);
     [min_ver("1.12")] string AccessRight;
     [min_ver("1.14")] ESParray<ESPstruct ServerJobQueue> ServerJobQueues;
+    [min_ver("1.22")] string ActivityTime;
+    [min_ver("1.22")] bool DaliDetached;
 };
 
 ESPrequest SMCIndexRequest
@@ -399,7 +401,7 @@ ESPresponse [exceptions_inline] LockQueryResponse
     int NumLocks;
 };
 
-ESPservice [auth_feature("DEFERRED"), noforms, version("1.21"), exceptions_inline("./smc_xslt/exceptions.xslt"), use_method_name] WsSMC
+ESPservice [auth_feature("DEFERRED"), noforms, version("1.22"), exceptions_inline("./smc_xslt/exceptions.xslt"), use_method_name] WsSMC
 {
     ESPmethod Index(SMCIndexRequest, SMCIndexResponse);
     ESPmethod [resp_xsl_default("/esp/xslt/index.xslt")] Activity(ActivityRequest, ActivityResponse);

--- a/esp/services/ws_smc/ws_smcService.cpp
+++ b/esp/services/ws_smc/ws_smcService.cpp
@@ -55,7 +55,6 @@ static const char* SMC_ACCESS_DENIED = "Access Denied";
 static const char* QUEUE_ACCESS_DENIED = "Failed to access the queue functions. Permission denied.";
 
 const char* PERMISSIONS_FILENAME = "espsmc_permissions.xml";
-const unsigned DEFAULTACTIVITYINFOCACHETIMEOUTSECOND = 10;
 
 void AccessSuccess(IEspContext& context, char const * msg,...) __attribute__((format(printf, 2, 3)));
 void AccessSuccess(IEspContext& context, char const * msg,...)
@@ -142,10 +141,14 @@ void CWsSMCEx::init(IPropertyTree *cfg, const char *process, const char *service
         m_PortalURL.append(portalURL);
 
     xpath.setf("Software/EspProcess[@name=\"%s\"]/EspService[@name=\"%s\"]/ActivityInfoCacheSeconds", process, service);
-    activityInfoCacheSeconds = cfg->getPropInt(xpath.str(), DEFAULTACTIVITYINFOCACHETIMEOUTSECOND);
+    unsigned activityInfoCacheSeconds = cfg->getPropInt(xpath.str(), DEFAULTACTIVITYINFOCACHEFORCEBUILDSECOND);
     xpath.setf("Software/EspProcess[@name=\"%s\"]/EspService[@name=\"%s\"]/LogDaliConnection", process, service);
     if (cfg->getPropBool(xpath.str()))
         querySDS().setConfigOpt("Client/@LogConnection", "true");
+
+    xpath.setf("Software/EspProcess[@name=\"%s\"]/EspService[@name=\"%s\"]/ActivityInfoCacheAutoRebuildSeconds", process, service);
+    unsigned activityInfoCacheAutoRebuildSeconds = cfg->getPropInt(xpath.str(), DEFAULTACTIVITYINFOCACHEAUTOREBUILDSECOND);
+    activityInfoReader.setown(new CActivityInfoReader(activityInfoCacheAutoRebuildSeconds, activityInfoCacheSeconds));
 }
 
 struct CActiveWorkunitWrapper: public CActiveWorkunit
@@ -1194,7 +1197,9 @@ bool CWsSMCEx::onActivity(IEspContext &context, IEspActivityRequest &req, IEspAc
         if (version >= 1.06)
             setBannerAndChatData(version, resp);
 
-        Owned<CActivityInfo> activityInfo = getActivityInfo(context);
+        Owned<CActivityInfo> activityInfo = activityInfoReader->getActivityInfo();
+        if (!activityInfo)
+            throw MakeStringException(ECLWATCH_INTERNAL_ERROR, "Failed to get Activity Info. Please try later.");
         setActivityResponse(context, activityInfo, req, resp);
     }
     catch(IException* e)
@@ -1203,29 +1208,6 @@ bool CWsSMCEx::onActivity(IEspContext &context, IEspActivityRequest &req, IEspAc
     }
 
     return true;
-}
-
-void CWsSMCEx::clearActivityInfoCache()
-{
-    CriticalBlock b(getActivityCrit);
-    activityInfoCache.clear();
-}
-
-CActivityInfo* CWsSMCEx::getActivityInfo(IEspContext &context)
-{
-    CriticalBlock b(getActivityCrit);
-
-    if (activityInfoCache && activityInfoCache->isCachedActivityInfoValid(activityInfoCacheSeconds))
-        return activityInfoCache.getLink();
-
-    DBGLOG("CWsSMCEx::getActivityInfo - rebuild cached information");
-    {
-        EspTimeSection timer("createActivityInfo");
-        activityInfoCache.setown(new CActivityInfo());
-        activityInfoCache->createActivityInfo(context);
-    }
-
-    return activityInfoCache.getLink();
 }
 
 void CWsSMCEx::addWUsToResponse(IEspContext &context, const IArrayOf<IEspActiveWorkunit>& aws, IEspActivityResponse& resp)
@@ -1279,6 +1261,12 @@ void CWsSMCEx::setActivityResponse(IEspContext &context, CActivityInfo* activity
     double version = context.getClientVersion();
     const char* sortBy = req.getSortBy();
     bool descending = req.getDescending();
+    if (version >= 1.22)
+    {
+        StringBuffer s;
+        resp.setActivityTime(activityInfo->queryTimeCached(s));
+        resp.setDaliDetached(activityInfoReader->isDaliDetached());
+    }
     if (version >= 1.16)
     {
         IArrayOf<IEspTargetCluster> thorClusters;
@@ -1419,7 +1407,7 @@ bool CWsSMCEx::onMoveJobDown(IEspContext &context, IEspSMCJobRequest &req, IEspS
             }
         }
         AccessSuccess(context, "Changed job priority %s",req.getWuid());
-        clearActivityInfoCache();
+        activityInfoReader->rebuild();
         resp.setRedirectUrl("/WsSMC/");
     }
     catch(IException* e)
@@ -1448,7 +1436,7 @@ bool CWsSMCEx::onMoveJobUp(IEspContext &context, IEspSMCJobRequest &req, IEspSMC
             }
         }
         AccessSuccess(context, "Changed job priority %s",req.getWuid());
-        clearActivityInfoCache();
+        activityInfoReader->rebuild();
         resp.setRedirectUrl("/WsSMC/");
     }
     catch(IException* e)
@@ -1494,7 +1482,7 @@ bool CWsSMCEx::onMoveJobBack(IEspContext &context, IEspSMCJobRequest &req, IEspS
             }
         }
         AccessSuccess(context, "Changed job priority %s",req.getWuid());
-        clearActivityInfoCache();
+        activityInfoReader->rebuild();
         resp.setRedirectUrl("/WsSMC/");
     }
     catch(IException* e)
@@ -1541,7 +1529,7 @@ bool CWsSMCEx::onMoveJobFront(IEspContext &context, IEspSMCJobRequest &req, IEsp
         }
 
         AccessSuccess(context, "Changed job priority %s",req.getWuid());
-        clearActivityInfoCache();
+        activityInfoReader->rebuild();
         resp.setRedirectUrl("/WsSMC/");
     }
     catch(IException* e)
@@ -1571,7 +1559,7 @@ bool CWsSMCEx::onRemoveJob(IEspContext &context, IEspSMCJobRequest &req, IEspSMC
             }
         }
         AccessSuccess(context, "Removed job %s",req.getWuid());
-        clearActivityInfoCache();
+        activityInfoReader->rebuild();
         resp.setRedirectUrl("/WsSMC/");
     }
     catch(IException* e)
@@ -1593,7 +1581,7 @@ bool CWsSMCEx::onStopQueue(IEspContext &context, IEspSMCQueueRequest &req, IEspS
             queue->stop(createQueueActionInfo(context, "stopped", req, info));
         }
         AccessSuccess(context, "Stopped queue %s", req.getCluster());
-        clearActivityInfoCache();
+        activityInfoReader->rebuild();
         double version = context.getClientVersion();
         if (version >= 1.19)
             getStatusServerInfo(context, req.getServerType(), req.getCluster(), req.getNetworkAddress(), req.getPort(), resp.updateStatusServerInfo());
@@ -1619,7 +1607,7 @@ bool CWsSMCEx::onResumeQueue(IEspContext &context, IEspSMCQueueRequest &req, IEs
             queue->resume(createQueueActionInfo(context, "resumed", req, info));
         }
         AccessSuccess(context, "Resumed queue %s", req.getCluster());
-        clearActivityInfoCache();
+        activityInfoReader->rebuild();
         double version = context.getClientVersion();
         if (version >= 1.19)
             getStatusServerInfo(context, req.getServerType(), req.getCluster(), req.getNetworkAddress(), req.getPort(), resp.updateStatusServerInfo());
@@ -1662,7 +1650,7 @@ bool CWsSMCEx::onPauseQueue(IEspContext &context, IEspSMCQueueRequest &req, IEsp
             queue->pause(createQueueActionInfo(context, "paused", req, info));
         }
         AccessSuccess(context, "Paused queue %s", req.getCluster());
-        clearActivityInfoCache();
+        activityInfoReader->rebuild();
         double version = context.getClientVersion();
         if (version >= 1.19)
             getStatusServerInfo(context, req.getServerType(), req.getCluster(), req.getNetworkAddress(), req.getPort(), resp.updateStatusServerInfo());
@@ -1692,7 +1680,7 @@ bool CWsSMCEx::onClearQueue(IEspContext &context, IEspSMCQueueRequest &req, IEsp
             queue->clear();
         }
         AccessSuccess(context, "Cleared queue %s",req.getCluster());
-        clearActivityInfoCache();
+        activityInfoReader->rebuild();
         double version = context.getClientVersion();
         if (version >= 1.19)
             getStatusServerInfo(context, req.getServerType(), req.getCluster(), req.getNetworkAddress(), req.getPort(), resp.updateStatusServerInfo());
@@ -1761,7 +1749,7 @@ bool CWsSMCEx::onSetJobPriority(IEspContext &context, IEspSMCPriorityRequest &re
             }
         }
 
-        clearActivityInfoCache();
+        activityInfoReader->rebuild();
         resp.setRedirectUrl("/WsSMC/");
     }
     catch(IException* e)
@@ -2255,9 +2243,9 @@ void CWsSMCEx::getStatusServerInfo(IEspContext &context, const char *serverType,
     if (!serverType || !*serverType)
         throw MakeStringException(ECLWATCH_MISSING_PARAMS, "Server type not specified.");
 
-    Owned<CActivityInfo> activityInfo = getActivityInfo(context);
+    Owned<CActivityInfo> activityInfo = activityInfoReader->getActivityInfo();
     if (!activityInfo)
-        throw MakeStringException(ECLWATCH_INTERNAL_ERROR, "Failed to get Activity Info cache.");
+        throw MakeStringException(ECLWATCH_INTERNAL_ERROR, "Failed to get Activity Info. Please try later.");
 
     if (strieq(serverType,STATUS_SERVER_THOR))
     {
@@ -2621,3 +2609,53 @@ bool CWsSMCEx::onLockQuery(IEspContext &context, IEspLockQueryRequest &req, IEsp
     return true;
 }
 
+void CActivityInfoReader::threadmain()
+{
+    PROGLOG("WsSMC CActivityInfoReader Thread started.");
+    unsigned int autoRebuildMillSeconds = 1000*autoRebuildSeconds;
+    while (!stopping)
+    {
+        if (!detached)
+        {
+            try
+            {
+                EspTimeSection timer("createActivityInfo");
+                Owned<IEspContext> espContext =  createEspContext();
+                Owned<CActivityInfo> activityInfo = new CActivityInfo();
+                activityInfo->createActivityInfo(*espContext);
+                PROGLOG("WsSMC CActivityInfoReader: ActivityInfo collected.");
+
+                CriticalBlock b(crit);
+                activityInfoCache.setown(activityInfo.getClear());
+
+                // if 1st and getActivityInfo blocked, release it.
+                if (first)
+                {
+                    first = false;
+                    if (firstBlocked)
+                    {
+                        firstBlocked = false;
+                        firstSem.signal();
+                    }
+                }
+            }
+            catch(IException *e)
+            {
+                StringBuffer msg;
+                IERRLOG("Exception %d:%s in WsSMC CActivityInfoReader::run", e->errorCode(), e->errorMessage(msg).str());
+                e->Release();
+            }
+            catch(...)
+            {
+                IERRLOG("Unknown exception in WsSMC CActivityInfoReader::run");
+            }
+        }
+
+        waiting = true;
+        if (!sem.wait(autoRebuildMillSeconds))
+        {
+            bool expected = true;
+            waiting.compare_exchange_strong(expected, false);
+        }
+    }
+}

--- a/esp/services/ws_smc/ws_smcService.hpp
+++ b/esp/services/ws_smc/ws_smcService.hpp
@@ -24,6 +24,9 @@
 #include "TpWrapper.hpp"
 #include "WUXMLInfo.hpp"
 
+const unsigned DEFAULTACTIVITYINFOCACHEFORCEBUILDSECOND = 10;
+const unsigned DEFAULTACTIVITYINFOCACHEAUTOREBUILDSECOND = 120;
+
 enum BulletType
 {
     bulletNONE = 0,
@@ -154,15 +157,94 @@ public:
     inline IArrayOf<IEspActiveWorkunit>& queryActiveWUs() { return aws; };
     inline IArrayOf<IEspServerJobQueue>& queryServerJobQueues() { return serverJobQueues; };
     inline IArrayOf<IEspDFUJob>& queryDFURecoveryJobs() { return DFURecoveryJobs; };
+    inline StringBuffer& queryTimeCached(StringBuffer& str) { return timeCached.getString(str, true); }
+};
+
+class CWsSMCEx;
+
+class CActivityInfoReader : public CInterface, implements IThreaded
+{
+    bool stopping = false;
+    bool detached = false;
+    bool first = true;
+    bool firstBlocked = false;
+    unsigned autoRebuildSeconds = DEFAULTACTIVITYINFOCACHEAUTOREBUILDSECOND;
+    unsigned forceRebuildSeconds = DEFAULTACTIVITYINFOCACHEFORCEBUILDSECOND;
+    Owned<CActivityInfo> activityInfoCache;
+    Semaphore sem;
+    Semaphore firstSem;
+    CriticalSection crit;
+    CThreaded threaded;
+    std::atomic<bool> waiting = {false};
+public:
+    CActivityInfoReader(unsigned _autoRebuildSeconds, unsigned _forceRebuildSeconds)
+        : autoRebuildSeconds(_autoRebuildSeconds), forceRebuildSeconds(_forceRebuildSeconds), threaded("ActivityInfoReader")
+    {
+        threaded.init(this);
+    };
+
+    ~CActivityInfoReader()
+    {
+        stopping = true;
+        sem.signal();
+        firstSem.signal(); // in case
+        threaded.join();
+    }
+
+    virtual void threadmain() override;
+    CActivityInfo* getActivityInfo()
+    {
+        CLeavableCriticalBlock b(crit);
+        if (first)
+        {
+            if (detached)
+                return nullptr;
+            firstBlocked = true;
+            b.leave();
+            firstSem.wait();
+            b.enter();
+            if (first)
+                return nullptr;
+        }
+
+        //Now, activityInfoCache should always be available.
+        assertex(activityInfoCache);
+        if (!detached && !activityInfoCache->isCachedActivityInfoValid(forceRebuildSeconds))
+            rebuild();
+        return activityInfoCache.getLink();
+    }
+    void rebuild()
+    {
+        bool expected = true;
+        if (waiting.compare_exchange_strong(expected, false))
+            sem.signal();
+    }
+    void setDetachedState(bool _detached)
+    {
+        CriticalBlock b(crit);
+        if (detached != _detached)
+        {
+            detached = _detached;
+            if (detached)
+            {
+                // NB: first will still be true, signal and getActivityInfo() will return null
+                if (firstBlocked)
+                {
+                    firstBlocked = false;
+                    firstSem.signal(); // NB: first still true
+                }
+            }
+            else
+                rebuild();
+        }
+    }
+    bool isDaliDetached() const { return detached; }
 };
 
 class CWsSMCEx : public CWsSMC
 {
     long m_counter;
     CTpWrapper m_ClusterStatus;
-    CriticalSection getActivityCrit;
-    Owned<CActivityInfo> activityInfoCache;
-    unsigned activityInfoCacheSeconds;
 
     StringBuffer m_ChatURL;
     StringBuffer m_Banner;
@@ -173,6 +255,7 @@ class CWsSMCEx : public CWsSMC
     int m_BannerAction;
     bool m_EnableChatURL;
     CriticalSection crit;
+    Owned<CActivityInfoReader> activityInfoReader;
 
 public:
     IMPLEMENT_IINTERFACE;
@@ -195,6 +278,18 @@ public:
     bool onGetThorQueueAvailability(IEspContext &context, IEspGetThorQueueAvailabilityRequest &req, IEspGetThorQueueAvailabilityResponse& resp);
     bool onSetBanner(IEspContext &context, IEspSetBannerRequest &req, IEspSetBannerResponse& resp);
     bool onNotInCommunityEdition(IEspContext &context, IEspNotInCommunityEditionRequest &req, IEspNotInCommunityEditionResponse &resp);
+
+    virtual bool attachServiceToDali() override
+    {
+        activityInfoReader->setDetachedState(false);
+        return true;
+    }
+
+    virtual bool detachServiceFromDali() override
+    {
+        activityInfoReader->setDetachedState(true);
+        return true;
+    }
 
     virtual bool onBrowseResources(IEspContext &context, IEspBrowseResourcesRequest & req, IEspBrowseResourcesResponse & resp);
     virtual bool onRoxieControlCmd(IEspContext &context, IEspRoxieControlCmdRequest &req, IEspRoxieControlCmdResponse &resp);
@@ -221,8 +316,6 @@ private:
     void setJobPriority(IEspContext &context, IWorkUnitFactory* factory, const char* wuid, const char* queue, WUPriorityClass& priority);
 
     void setESPTargetClusters(IEspContext& context, const CIArrayOf<CWsSMCTargetCluster>& targetClusters, IArrayOf<IEspTargetCluster>& respTargetClusters);
-    void clearActivityInfoCache();
-    CActivityInfo* getActivityInfo(IEspContext &context);
     void setActivityResponse(IEspContext &context, CActivityInfo* activityInfo, IEspActivityRequest &req, IEspActivityResponse& resp);
     void addWUsToResponse(IEspContext &context, const IArrayOf<IEspActiveWorkunit>& aws, IEspActivityResponse& resp);
 

--- a/initfiles/componentfiles/configxml/@temp/esp_service_WsSMC.xsl
+++ b/initfiles/componentfiles/configxml/@temp/esp_service_WsSMC.xsl
@@ -153,6 +153,9 @@ This is required by its binding with ESP service '<xsl:value-of select="$espServ
             <xsl:if test="string(@ActivityInfoCacheSeconds) != ''">
                 <ActivityInfoCacheSeconds><xsl:value-of select="@ActivityInfoCacheSeconds"/></ActivityInfoCacheSeconds>
             </xsl:if>
+            <xsl:if test="string(@ActivityInfoCacheAutoRebuildSeconds) != ''">
+                <ActivityInfoCacheAutoRebuildSeconds><xsl:value-of select="@ActivityInfoCacheAutoRebuildSeconds"/></ActivityInfoCacheAutoRebuildSeconds>
+            </xsl:if>
             <xsl:if test="string(@enableLogDaliConnection) != ''">
                 <LogDaliConnection><xsl:value-of select="@enableLogDaliConnection"/></LogDaliConnection>
             </xsl:if>

--- a/initfiles/componentfiles/configxml/espsmcservice.xsd.in
+++ b/initfiles/componentfiles/configxml/espsmcservice.xsd.in
@@ -106,7 +106,14 @@
             <xs:attribute name="ActivityInfoCacheSeconds" type="xs:nonNegativeInteger" use="optional" default="10">
                 <xs:annotation>
                     <xs:appinfo>
-                        <tooltip>timeout for activity info cache (in seconds).</tooltip>
+                        <tooltip>If user makes a request and the cache is older than this value (in seconds), ESP returns the cache and starts to rebuild a new cache.</tooltip>
+                    </xs:appinfo>
+                </xs:annotation>
+            </xs:attribute>
+            <xs:attribute name="ActivityInfoCacheAutoRebuildSeconds" type="xs:nonNegativeInteger" use="optional" default="120">
+                <xs:annotation>
+                    <xs:appinfo>
+                        <tooltip>When idle, ESP automatically rebuilds the cache if this value (in seconds) exceeded.</tooltip>
                     </xs:appinfo>
                 </xs:annotation>
             </xs:attribute>

--- a/initfiles/etc/DIR_NAME/environment.xml.in
+++ b/initfiles/etc/DIR_NAME/environment.xml.in
@@ -623,6 +623,7 @@
              port="8877"/>
   </SashaServerProcess>
   <EspService ActivityInfoCacheSeconds="10"
+              ActivityInfoCacheAutoRebuildSeconds="120"
               allowNewRoxieOnDemandQuery="false"
               AWUsCacheTimeout="15"
               build="_"

--- a/testing/regress/environment.xml.in
+++ b/testing/regress/environment.xml.in
@@ -536,6 +536,7 @@
              port="8877"/>
   </SashaServerProcess>
   <EspService ActivityInfoCacheSeconds="10"
+              ActivityInfoCacheAutoRebuildSeconds="120"
               allowNewRoxieOnDemandQuery="false"
               AWUsCacheTimeout="15"
               build="_"

--- a/testing/unittests/configmgr/schema/environments/hpcc_port_conflict_test1.xml
+++ b/testing/unittests/configmgr/schema/environments/hpcc_port_conflict_test1.xml
@@ -34,6 +34,7 @@
     </Programs>
     <Software>
         <EspService ActivityInfoCacheSeconds="10"
+                    ActivityInfoCacheAutoRebuildSeconds="120"
                     allowNewRoxieOnDemandQuery="false"
                     AWUsCacheTimeout="15"
                     build="_"

--- a/testing/unittests/configmgr/schema/environments/hpcc_port_conflict_test2.xml
+++ b/testing/unittests/configmgr/schema/environments/hpcc_port_conflict_test2.xml
@@ -34,6 +34,7 @@
     </Programs>
     <Software>
         <EspService ActivityInfoCacheSeconds="10"
+                    ActivityInfoCacheAutoRebuildSeconds="120"
                     allowNewRoxieOnDemandQuery="false"
                     AWUsCacheTimeout="15"
                     build="_"

--- a/testing/unittests/configmgr/schema/environments/hpcc_port_conflict_test3.xml
+++ b/testing/unittests/configmgr/schema/environments/hpcc_port_conflict_test3.xml
@@ -34,6 +34,7 @@
     </Programs>
     <Software>
         <EspService ActivityInfoCacheSeconds="10"
+                    ActivityInfoCacheAutoRebuildSeconds="120"
                     allowNewRoxieOnDemandQuery="false"
                     AWUsCacheTimeout="15"
                     build="_"

--- a/testing/unittests/configmgr/schema/environments/hpcc_port_conflict_test4.xml
+++ b/testing/unittests/configmgr/schema/environments/hpcc_port_conflict_test4.xml
@@ -34,6 +34,7 @@
     </Programs>
     <Software>
         <EspService ActivityInfoCacheSeconds="10"
+                    ActivityInfoCacheAutoRebuildSeconds="120"
                     allowNewRoxieOnDemandQuery="false"
                     AWUsCacheTimeout="15"
                     build="_"

